iDiff/utils/image_utils.go

```diff
@@ -2,47 +2,132 @@ package utils
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
 	"io"
 	"os"
 	"path/filepath"
+	"regexp"
 	"strings"
 
+	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/client"
 	"github.com/docker/docker/pkg/system"
 )
 
-// ImageToDir converts an image to an unpacked tar and creates a representation of that directory.
-func ImageToDir(img string) (string, string, error) {
+func saveImageToTar(image string) (string, error) {
 	cli, err := client.NewEnvClient()
 	if err != nil {
-		return "", "", err
+		return "", err
 	}
-	tarPath, err := ImageToTar(cli, img)
+
+	fromImage := image
+	toTar := image
+	// If not an already existing image ID, have to pull it from a repo before saving it
+	if !checkImageID(image) {
+		imageID, imageName, err := pullImageFromRepo(cli, image)
+		if err != nil {
+			return "", err
+		}
+		fromImage = imageID
+		toTar = imageName
+	}
+	// Convert the image into a tar
+	imageTarPath, err := ImageToTar(cli, fromImage, toTar)
 	if err != nil {
-		return "", "", err
+		return "", err
+	}
+	return imageTarPath, nil
+}
+
+// ImageToDir converts an image to an unpacked tar and creates a representation of that directory.
+func ImageToDir(img string) (string, string, error) {
+	var tarName string
+	if !checkImageTar(img) {
+		// If not an image tar already existing in the filesystem, create client to obtain image
+		imageTar, err := saveImageToTar(img)
+		if err != nil {
+			return "", "", err
+		}
+		tarName = imageTar
+	} else {
+		tarName = img
+	}
+	defer os.Remove(tarName)
+	return TarToJSON(tarName)
+}
+
+type Event struct {
+	Status         string `json:"status"`
+	Error          string `json:"error"`
+	Progress       string `json:"progress"`
+	ProgressDetail struct {
+		Current int `json:"current"`
+		Total   int `json:"total"`
+	} `json:"progressDetail"`
+}
+
+func getImagePullResponse(image string, response []Event) (string, error) {
+	var imageDigest string
+	for _, event := range response {
+		if event.Error != "" {
+			err := fmt.Errorf("Error pulling image %s: %s", image, event.Error)
+			return "", err
+		}
+		digestPattern := regexp.MustCompile("^Digest: (sha256:[a-z|0-9]{64})$")
+		digestMatch := digestPattern.FindStringSubmatch(event.Status)
+		if len(digestMatch) != 0 {
+			imageDigest = digestMatch[1]
+			return imageDigest, nil
+		}
 	}
-	err = ExtractTar(tarPath)
+	err := fmt.Errorf("Could not pull image %s", image)
+	return "", err
+}
+
+func pullImageFromRepo(cli client.APIClient, image string) (string, string, error) {
+	response, err := cli.ImagePull(context.Background(), image, types.ImagePullOptions{})
```

Comment:
It is important to note that this may not work properly for all users.  This will invoke docker client commands(from the lib) to a docker daemon(pre-installed on the user's host machine) and the client/daemon might not have the correct version compatibility.  I believe Abby is working on a PR to resolve this but just going to note it here.